### PR TITLE
Add PC_Affine for patches and points

### PR DESCRIPTION
**Pull request by elemoine**
This PR adds `PC_Affine(pcpatch)` and `PC_Affine(pcpoint)` for applying affine transformations to patches and points, respectively.

### CODE REVIEW

File: lib/cunit/cu_pc_point.c [hunk 1]
```diff
@@ -220,12 +220,51 @@ test_point_rotate_quaternion()
...
+    pc_point_free(pt);
+}
+
 /* REGISTER ***********************************************************/
 
 CU_TestInfo point_tests[] = {
 	PC_TEST(test_point_hex_inout),
 	PC_TEST(test_point_access),
     PC_TEST(test_point_rotate_quaternion),
+    PC_TEST(test_point_affine),
```
Comment (mbredif) [inline lib/cunit/cu_pc_point.c]: indentation

File: lib/pc_api_internal.h [hunk 2]
```diff
@@ -90,6 +90,8 @@ typedef struct
 /* */
 typedef double PCMAT33[9];
 typedef double PCVEC3[3];
+typedef double PCMAT43[12];
+typedef double PCVEC4[4];
```
Comment (mbredif) [inline lib/pc_api_internal.h]: unused for now

File: lib/pc_api_internal.h [hunk 3]
```diff
@@ -299,7 +301,9 @@ void pc_bitmap_filter(PCBITMAP *map, PC_FILTERTYPE filter, int i, double d, doub
 */
 
 void pc_matrix_set_from_quaternion(PCMAT33 mat, double qw, double qx, double qy, double qz);
+void pc_matrix_set_affine(PCMAT43 mat, double a, double b, double c, double d, double e, double f, double g, double h, double i, double xoff, double yoff, double zoff);
 void pc_matrix_multiply_vector(PCVEC3 rotatedvec, const PCMAT33 mat, const PCVEC3 vec);
+void pc_matrix_multiply_vector_43(PCVEC3 res, const PCMAT43 mat, const PCVEC4 vec);
```
Comment (mbredif) [inline lib/pc_api_internal.h]: vec could be a PCVEC3

File: lib/pc_matrix.c [hunk 4]
```diff
@@ -61,3 +84,18 @@ pc_matrix_multiply_vector(PCVEC3 res, const PCMAT33 mat, const PCVEC3 vec)
...
+void
+pc_matrix_multiply_vector_43(PCVEC3 res, const PCMAT43 mat, const PCVEC4 vec)
+{
+    res[0] = mat[0] * vec[0] + mat[1] * vec[1]
+        + mat[2] * vec[2] + mat[3] * vec[3];
+    res[1] = mat[4] * vec[0] + mat[5] * vec[1]
+        + mat[6] * vec[2] + mat[7] * vec[3];
+    res[2] = mat[8] * vec[0] + mat[9] * vec[1]
+        + mat[10] * vec[2] + mat[11] * vec[3];
+}
```
Comment (mbredif) [inline lib/pc_matrix.c]: vec could be a PCVEC3, indexing issues for mat, multiplication by vec[3]==1 should not be performed

**Review by mbredif — CHANGES_REQUESTED** [review]
Sounds good as well. The maths are wrong though

File: lib/cunit/cu_pc_point.c (hunk 1 above)
Comment (elemoine) [reply lib/cunit/cu_pc_point.c]: My indent is correct. The one for `test_point_hex_inout` and `test_point_access` is not. Fixing the indent in this file should be done with a separate change.

File: lib/pc_api_internal.h (hunk 2 above)
Comment (elemoine) [reply lib/pc_api_internal.h]: It is used. I need a 4-dimension vector for the 4x3 matrix multiplication.

File: lib/pc_api_internal.h (hunk 3 above)
Comment (elemoine) [reply lib/pc_api_internal.h]: No, I don't think so. It needs to be a 4-dimension vector with the last coordinate set to 1.

**Comment by elemoine** [comment]
> Sounds good as well. The maths are wrong though

Sorry, I don't understand what's wrong with the math. Please pinpoint the problem. Thanks.

**Comment by mbredif** [comment]
My mistake, the maths are good. It is just that you can save 3 multiplies by using a PCVEC3 and not multiplying by 1.
`res[2] = mat[8] * vec[0] + mat[9] * vec[1] + mat[10] * vec[2] + mat[11];`
Apart from that, this can be merged.

**Comment by elemoine** [comment]
> My mistake, the maths are good. It is just that you can save 3 multiplies by using a PCVEC3 and not multiplying by 1.
> res[2] = mat[8] * vec[0] + mat[9] * vec[1] + mat[10] * vec[2] + mat[11];
> Apart from that, this can be merged.

Yep, I'll finish that up tomorrow. Thanks for your reviews.

**Comment by elemoine** [comment]
Done. I removed `PCVEC4`. Hope this matches what you had in mind @mbredif.

File: pgsql/pc_editor.c [hunk 5]
```diff
@@ -111,79 +226,87 @@ Datum pcpoint_rotate_quaternion(PG_FUNCTION_ARGS)
 }
 
 /**
-* Translate a patch
-* PC_Translate(patch pcpatch, tx float8, ty float8, tz float8,
-*              xdimname text, ydimname text, zdimname text) returns pcpatch
+* Translate a point
+* PC_RotateTranslate(point pcpoint, tx float8, ty float8, tz float8,
```
Comment (mbredif) [inline pgsql/pc_editor.c]: PC_RotateTranslate ??

**Review by mbredif — APPROVED** [review]
Good to go (except the comment issue with PC_RotateTranslate)

File: pgsql/pc_editor.c (hunk 5 above)
Comment (elemoine) [reply pgsql/pc_editor.c]: The mistake was there already. I'll fix with a new commit…

**Comment by elemoine** [comment]
> Good to go (except the comment issue with PC_RotateTranslate)

I added a commit to fix that typo. I was not introduced by this PR's initial commits.

Thanks a lot for the sharp eyes :)

**Comment by mbredif** [comment]
Thanks @elemoine !

**Pull request merged**